File: 567-permutation-in-string/567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        
        unordered_map<char,int> hash;
        
        for(int i = 0; i < s1.size(); i++) {
            hash[s1[i]]++;
        }
        
        int matches = 0;
        
        int left = 0;
        
        for(int r = 0; r < s2.size(); r++) {
            char curChar = s2[r];
            
            if(hash.count(curChar)) {
                hash[curChar]--;
                if(hash[curChar] == 0) {
                    matches++;
                }
            }
            
            if(matches == hash.size()) {
                return true;
            }
            
            if(r >= s1.size()-1) {
                char leftChar = s2[left];
                if(hash.count(leftChar)) {
                    if(hash[leftChar] == 0) {
                        matches--;
                    }
                    hash[leftChar]++;
                }
                left++;
            }
        }
        
        return false;
        
    }
};

File: 567-permutation-in-string/567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        
        left = 0
        matches = 0
        
        ht = {}
        
        for i in range(len(s1)):
            if s1[i] not in ht:
                ht[s1[i]] = 1
            else:
                ht[s1[i]]+=1
        
        for r in range(len(s2)):
            curChar = s2[r]
            
            if curChar in ht:
                ht[curChar]-=1
                if ht[curChar] == 0:
                    matches+=1
            
            if(matches == len(ht)):
                return True
            
            if(r >= len(s1) - 1):
                leftChar = s2[left]
                if leftChar in ht:
                    if ht[leftChar] == 0:
                        matches-=1
                    ht[leftChar]+=1
                left+=1
                
            
        return False
        